### PR TITLE
Update AND and OR operation evaluation to properly propagate errors

### PR DESCRIPTION
**Pull request by clemire**
Logical AND and ORs were dropping errors and causing false negatives for entitlement evaluations. This PR makes changes Logical operand evaluation to ignore errors only when we definitely know the user was or was not entitled.

specifically, for AND operations:
- if a single child evaluates to false, return false
- otherwise, if either errors, return an error

For OR operations:
- if a single child evaluates to true, return true
- otherwise, if either child operation errors, return an error

### CODE REVIEW

File: core/xchain/entitlement/check_operation.go [hunk 1]
```diff
@@ -179,11 +179,14 @@ func (e *Evaluator) evaluateMockOperation(
...
-	if op.ChainID.Sign() != 0 {
-		return true, nil
-	} else {
-		return false, nil
+
+	if (op.ContractAddress != common.Address{}) {
+		// Grab last two digits of contract address as a unique identifier of which check
+		// caused the error, for ease of debugging test cases.
+		code := fmt.Sprintf("%v", op.ContractAddress)[40:]
+		return false, fmt.Errorf("intentional failure (%v)", code)
```
Comment (bas-vk) [inline core/xchain/entitlement/check_operation.go:187]: You can also do `fmt.Errorf("intentional failure (%x%x)", op.ContractAddress[18], op.ContractAddress[19])` to save on mem allocations.

File: core/xchain/entitlement/entitlement.go [hunk 2]
```diff
@@ -26,6 +26,21 @@ func (e *Evaluator) EvaluateRuleData(
 	return e.evaluateOp(ctx, opTree, linkedWallets)
 }
 
+func isNilOrCancelled(err error) bool {
+	return err == nil || err == context.Canceled
```
Comment (bas-vk) [inline core/xchain/entitlement/entitlement.go:30]: Consider using `errors.Is(err, context.Cancelled)`. This also verifies is err is wrapped around a canceled err which is probably what you want to check for.

File: core/xchain/entitlement/entitlement.go [hunk 3]
```diff
@@ -88,26 +130,39 @@ func (e *Evaluator) evaluateOrOperation(
...
 		}
 		wg.Done()
 	}()
 
 	wg.Wait()
-	return leftResult || rightResult, nil
+	// If at least one child evaluates as entitled, log any errors and return a true result.
+	if leftResult || rightResult {
+		return true, nil
+	} else {
```
Comment (bas-vk) [inline core/xchain/entitlement/entitlement.go:155]: To improve readability you can remove the else part and reduce nesting 1 lvl.

File: core/xchain/entitlement/entitlement.go [hunk 4]
```diff
@@ -88,26 +130,39 @@ func (e *Evaluator) evaluateOrOperation(
...
+		if leftErr != nil && rightErr != nil {
+			return false, fmt.Errorf("%w; %w", leftErr, rightErr)
+		} else {
+			finalErr := leftErr
+			if rightErr != nil {
+				finalErr = rightErr
+			}
+			return false, finalErr
```
Comment (bas-vk) [inline core/xchain/entitlement/entitlement.go:163]: The and operation logic checks if an error was the result of the context cancelled. Should that be checked here as well?

File: core/xchain/entitlement/entitlement.go (hunk 2 above)
Comment (clemire) [reply core/xchain/entitlement/entitlement.go:30]: Nice, thank you!

File: core/xchain/entitlement/entitlement.go (hunk 4 above)
Comment (clemire) [reply core/xchain/entitlement/entitlement.go:163]: Yes, I think so. Probably I just didn't catch this because it didn't occur when I was running tests.

**Review by texuf — APPROVED** [review]

**Pull request merged**